### PR TITLE
Add Room entity, Room manager (initial structure) and some Exceptions

### PR DESCRIPTION
**Pull request by Benjamin-Chain**
## Summary
- Created room entity
- Created room manager
- Created custom exceptions
- Unit testing over functionality

### CODE REVIEW

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.teamcubation.librarymanagement.domain.entities;
+
+import java.util.Objects;
+
+public class Room {
+    static int roomIds=0;
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java]: Please make the roomsIds variable private, there is no need to have this one as default.

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    private String address;
+    private int id;
+
+    public Room(String name, String address) {
+        this.name = name;
+        this.address = address;
+        this.id = Room.roomIds++;
+    }
+
+    public static int getRoomIds() {
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java]: Same here, there is no need to expose that variable outside.

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public String getAddress() {
+        return address;
+    }
+
+    public int getId() {
+        return id;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java:57]: Given that there is an Id for the rooms. wouldn't it be possible to use the id for the equals?

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.teamcubation.librarymanagement.domain.exceptions.room;
+
+public class RoomIncompleteFieldsException extends Exception{
+
+    public RoomIncompleteFieldsException(){
+        super("Some required fields are incomplete");
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java]: this error message can be a little bit more specific

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.teamcubation.librarymanagement.domain.exceptions.room;
+
+public class RoomIncompleteFieldsException extends Exception{
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java]: Please apply the formatter to the code, the braces should have a space before opening.

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.teamcubation.librarymanagement.domain.exceptions.room;
+
+public class RoomNotFoundException extends Exception{
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomNotFoundException.java]: same here it should be Exception {

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomNotFoundException.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+    public RoomNotFoundException(){
+        super("Room not found");
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomNotFoundException.java:6]: and here as well

File: src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.teamcubation.librarymanagement.domain.entities.Room;
+import com.teamcubation.librarymanagement.domain.entities.User;
+import com.teamcubation.librarymanagement.domain.exceptions.room.RoomIncompleteFieldsException;
+
+import java.util.ArrayList;
+import java.util.List;
+public class RoomManager {
+
+    private List<Room> rooms;
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java]: does rooms hold all the rooms or just the ones available? please name accordingly the variable.

File: src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    public boolean existRoom(Room room){
+        return rooms.contains(room) || reservedRooms.contains(room);
+    }
+
+    public boolean addRoom(Room room) throws RoomIncompleteFieldsException {
+        if(rooms.contains(room)){
+            return false;
+        }
+
+        if(room.getName() == null || room.getAddress() == null || room.getName().isEmpty() || room.getAddress().isEmpty()){
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java]: I think this error should arise directly from the creation of the room, (when doing the new) as it is the room responsibility to decide which fields are compulsory an which ones aren't.

File: src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    public boolean reserveRoom(Room room) throws RoomIncompleteFieldsException {
+        if(reservedRooms.contains(room)){
+            return false;
+        }
+
+        if(room.getName() == null || room.getAddress() == null){
+            throw new RoomIncompleteFieldsException();
+        }
+
+        reservedRooms.add(room);
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java]: adding a room here to the reserve rooms does not guarantee the room exists.

File: src/test/java/com/teamcubation/librarymanagement/domain/RoomServiceTest.java
Comment (gnievestm) [inline src/test/java/com/teamcubation/librarymanagement/domain/RoomServiceTest.java:1]: It looks like the main functionality created in this pr is over roomManager, I would expect to have test over it as well.

**Review by gnievestm — CHANGES_REQUESTED** [review]
Please check out the comments. Otherwise, great work.

**Comment by gnievestm** [comment]
Please add a brief decription of the work done in the pr description.

**Comment by gnievestm** [comment]
Curthermore name the branches as {action}/{jira-ticket}-{brief-description-kebab-case}

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java (hunk 1 above)
Comment (Benjamin-Chain) [reply src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java]: resolved

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java (hunk 2 above)
Comment (Benjamin-Chain) [reply src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java]: Thanks for the feedback

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java (hunk 3 above)
Comment (Benjamin-Chain) [reply src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java:57]: You have all reason, I think it's already resolved

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java (hunk 4 above)
Comment (Benjamin-Chain) [reply src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java]: I think it's already resolved and with a high level of specification

File: src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java (hunk 5 above)
Comment (Benjamin-Chain) [reply src/main/java/com/teamcubation/librarymanagement/domain/exceptions/room/RoomIncompleteFieldsException.java]: Okay, thanks

**Comment by Benjamin-Chain** [comment]
I think I have done all the corrections possible

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java [hunk 11]
```diff
@@ -1,21 +1,42 @@
...
         this.name = name;
         this.address = address;
-        this.id = Room.roomIds++;
+
+        if (id<=0) {
+            throw new RoomInvalidIdException();
+        } else {
+            this.id = id;
+        }
     }
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java:30]: There should be a blank line after finishing a methods for reading simplicity

File: src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java [hunk 12]
```diff
@@ -1,21 +1,42 @@
...
 
 public class Room {
-    static int roomIds=0;
+    private static int roomIds = 0;
     private String name;
     private String address;
     private int id;
 
-    public Room(String name, String address) {
+    public Room(int id, String name, String address) throws RoomIncompleteFieldsException, RoomInvalidIdException {
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java:14]: **I haven't check yet if it is done** -> given that the id is externally provided, the system should check if the id is unique in the manager before adding.

Comment (gnievestm) [reply src/main/java/com/teamcubation/librarymanagement/domain/entities/Room.java:14]: Found the check in the code :) well done.

File: src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java [hunk 13]
```diff
@@ -1,49 +1,53 @@
...
+        availableRooms.add(room);
         return true;
     }
 
-    public boolean reserveRoom(Room room) throws RoomIncompleteFieldsException {
+    public boolean reserveRoom(Room room) throws RoomIncompleteFieldsException, RoomNotFoundException, RoomAlreadyExistsException {
+
         if(reservedRooms.contains(room)){
-            return false;
+            throw new RoomAlreadyExistsException();
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java]: It seems like the exception here is that the room is already reserved, not that it already exists right?

File: src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java [hunk 14]
```diff
@@ -1,49 +1,53 @@
...
         }
 
         if(room.getName() == null || room.getAddress() == null){
             throw new RoomIncompleteFieldsException();
```
Comment (gnievestm) [inline src/main/java/com/teamcubation/librarymanagement/domain/managers/RoomManager.java]: There is no need to check if the room has the necessary fields, we can safely assume that only rooms that have already been added are avaialble for reserve, basically there are 2 possible cases, the room exist and it is reserved or it doesn't and an error is thrown. Regarding errors there should be 2 errors, room not found or room not available.

**Review by leosilva12 — APPROVED** [review]

**Pull request merged**